extensions/Lily/SoundExpanded.js

```diff
@@ -0,0 +1,365 @@
+      const sprite = util.target.sprite;
+
+      const soundId = sprite.sounds[index].soundId;
+      const soundPlayer = sprite.soundBank.soundPlayers[soundId];
+
+      if (!soundPlayer.outputNode) return false;
+      return soundPlayer.outputNode.loop;
+    }
+
+    stopSound(args, util) {
+      const index = this._getSoundIndex(args.SOUND, util);
+      if (index < 0) return 0;
+      const target = util.target;
+      const sprite = target.sprite;
+
+      const soundId = sprite.sounds[index].soundId;
+      const soundBank = sprite.soundBank;
+      soundBank.stop(target, soundId);
+    }
+
+    pauseSounds(args, util) {
+      this._toggleSoundState(args, util, true);
+    }
+
+    resumeSounds(args, util) {
+      this._toggleSoundState(args, util, false);
+    }
+
+    _toggleSoundState(args, util, state) {
+      const sprite = util.target.sprite;
+      const audioContext = sprite.soundBank.audioEngine.audioContext;
+
+      if (state) {
+        audioContext.suspend();
+        return;
+      } else {
+        audioContext.resume();
+        return;
+      }
+    }
+
+    isSoundPlaying(args, util) {
+      const index = this._getSoundIndex(args.SOUND, util);
+      if (index < 0) return false;
+      const sprite = util.target.sprite;
+
+      const soundId = sprite.sounds[index].soundId;
+      const soundPlayers = sprite.soundBank.soundPlayers;
+      return soundPlayers[soundId].isPlaying;
+    }
+
+    attributeOfSound(args, util) {
+      const index = this._getSoundIndex(args.SOUND, util);
+      if (index < 0) return 0;
+      const sprite = util.target.sprite;
+
+      const sound = sprite.sounds[index];
+      const soundId = sound.soundId;
+      const soundPlayer = sprite.soundBank.soundPlayers[soundId];
+      const soundBuffer = soundPlayer.buffer;
+
+      switch (args.ATTRIBUTE) {
+        case "length":
+          return Math.round(soundBuffer.duration * 100) / 100;
+        case "channels":
+          return soundBuffer.numberOfChannels;
+        case "sample rate":
+          return soundBuffer.sampleRate;
+        case "dataURI":
+          return sound.asset.encodeDataURI();
+      }
+    }
+
+    getSoundEffect(args, util) {
+      let target = Scratch.vm.runtime.getSpriteTargetByName(args.TARGET);
+      if (args.TARGET === "_myself_") target = util.target;
+      if (args.TARGET === "_stage_") target = runtime.getTargetForStage();
+      const effects = target.soundEffects;
+      if (!effects) return 0;
+      return effects[args.EFFECT];
+    }
+
+    setProjectVolume(args) {
+      const value = Scratch.Cast.toNumber(args.VALUE);
+      const newVolume = this._wrapClamp(value / 100, 0, 1);
+      runtime.audioEngine.inputNode.gain.value = newVolume;
+    }
+
+    changeProjectVolume(args) {
+      const value = Scratch.Cast.toNumber(args.VALUE) / 100;
+      const volume = runtime.audioEngine.inputNode.gain.value;
+      const newVolume = this._wrapClamp(volume + value, 0, 1);

```

Comment:
when volume is 0 and someone changes it by -1, it probably shouldn't wrap around to 199%

Reply:
Good point, will change that